src/libstd/sys/windows/stdio.rs

```diff
@@ -1,140 +1,225 @@
 #![unstable(issue = "0", feature = "windows_stdio")]
 
-use io::prelude::*;
-
+use cell::Cell;
 use cmp;
-use io::{self, Cursor};
+use io;
 use ptr;
 use str;
-use sync::Mutex;
 use sys::c;
 use sys::cvt;
 use sys::handle::Handle;
 
-pub enum Output {
-    Console(c::HANDLE),
-    Pipe(c::HANDLE),
-}
-
+// Don't cache handles but get them fresh for every read/write. This allows us to track changes to
+// the value over time (such as if a process calls `SetStdHandle` while it's running). See #40490.
 pub struct Stdin {
-    utf8: Mutex<io::Cursor<Vec<u8>>>,
+    high_surrogate: Cell<u16>,
 }
 pub struct Stdout;
 pub struct Stderr;
 
-pub fn get(handle: c::DWORD) -> io::Result<Output> {
-    let handle = unsafe { c::GetStdHandle(handle) };
+// Apparently Windows doesn't handle large reads on stdin or writes to stdout/stderr well (see
+// #13304 for details).
+//
+// From MSDN (2011): "The storage for this buffer is allocated from a shared heap for the
+// process that is 64 KB in size. The maximum size of the buffer will depend on heap usage."
+//
+// We choose the cap at 8 KiB because libuv does the same, and it seems to be acceptable so far.
+const MAX_BUFFER_SIZE: usize = 8196;
```

Comment:
Should this be 8192 to be power of 2?

Reply:
Why not use `1 << 13`? No typo then.

Reply:
Yes, a silly typo so thanks @iancormac84 for catching it. I can't read `1 << 13` and `0x2000` easy enough to directly see they are the same as 8 KiB.